connect/runtime/src/test/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperatorTest.java

```diff
@@ -199,112 +202,156 @@ private RetryWithToleranceOperator setupExecutor() {
         retryWithToleranceOperator.metrics(errorHandlingMetrics);
 
         EasyMock.expect(mockOperation.call()).andThrow(e).times(numRetriableExceptionsThrown);
         EasyMock.expect(mockOperation.call()).andReturn("Success");
+        for (Long expectedWait : expectedWaits) {
+            EasyMock.expect(exitLatch.await(expectedWait, TimeUnit.MILLISECONDS)).andAnswer(() -> {
+                time.sleep(expectedWait);
+                return false;
+            });
+        }
 
-        replay(mockOperation);
+        replay(mockOperation, exitLatch);
 
         String result = retryWithToleranceOperator.execAndHandleError(mockOperation, Exception.class);
-        assertTrue(retryWithToleranceOperator.failed());
-        assertNull(result);
-        assertEquals(expectedWait, time.hiResClockMs());
+
+        if (successExpected) {
+            assertFalse(retryWithToleranceOperator.failed());
+            assertEquals("Success", result);
+        } else {
+            assertTrue(retryWithToleranceOperator.failed());
+        }
 
         PowerMock.verifyAll();
     }
 
-    @Test
-    public void testCheckRetryLimit() {
+    public void execAndHandleNonRetriableError(int numRetriableExceptionsThrown, Exception e) throws Exception {
         MockTime time = new MockTime(0, 0, 0);
-        RetryWithToleranceOperator retryWithToleranceOperator = new RetryWithToleranceOperator(500, 100, NONE, time);
-
-        time.setCurrentTimeMs(100);
-        assertTrue(retryWithToleranceOperator.checkRetry(0));
+        CountDownLatch exitLatch = EasyMock.createStrictMock(CountDownLatch.class);
+        RetryWithToleranceOperator retryWithToleranceOperator = new RetryWithToleranceOperator(6000, ERRORS_RETRY_MAX_DELAY_DEFAULT, ALL, time, new ProcessingContext(), exitLatch);
+        retryWithToleranceOperator.metrics(errorHandlingMetrics);
 
-        time.setCurrentTimeMs(200);
-        assertTrue(retryWithToleranceOperator.checkRetry(0));
+        EasyMock.expect(mockOperation.call()).andThrow(e).times(numRetriableExceptionsThrown);
+        EasyMock.expect(mockOperation.call()).andReturn("Success");
 
-        time.setCurrentTimeMs(400);
-        assertTrue(retryWithToleranceOperator.checkRetry(0));
+        // expect no call to exitLatch.await() which is only called during the retry backoff
 
-        time.setCurrentTimeMs(499);
-        assertTrue(retryWithToleranceOperator.checkRetry(0));
+        replay(mockOperation, exitLatch);
 
-        time.setCurrentTimeMs(501);
-        assertFalse(retryWithToleranceOperator.checkRetry(0));
+        String result = retryWithToleranceOperator.execAndHandleError(mockOperation, Exception.class);
+        assertTrue(retryWithToleranceOperator.failed());
+        assertNull(result);
 
-        time.setCurrentTimeMs(600);
-        assertFalse(retryWithToleranceOperator.checkRetry(0));
+        PowerMock.verifyAll();
```

Comment:
I've added this verification to `execAndHandleRetriableError` (in the success case). But here in `execAndHandleNonRetriableError`, no matter what the value for `numNonRetriableExceptionsThrown` is, `mockOperation.call()` will only be called once (since the exception isn't retriable) - so we can't verify that the expectation that `e` is thrown `numNonRetriableExceptionsThrown` times is met. I guess this would've been easier to test if this was migrated to Mockito?

Reply:
If we don't actually invoke `mockOperation` more than once in any of the test cases with non-retriable errors, do we even need different test cases for 1 vs. 3 invocations? What's the difference between the two?

I also don't really follow why we're only verifying `mockOperation` in the success cases with retriable errors; what's preventing us from verifying it unconditionally?

Reply:
> do we even need different test cases for 1 vs. 3 invocations? What's the difference between the two?

There isn't any difference between the test cases for 1 vs 3 invocations for non-retriable errors, that's why I suggested consolidating them into a single test case in my previous comment (I've made this change now).

> I also don't really follow why we're only verifying mockOperation in the success cases with retriable errors; what's preventing us from verifying it unconditionally?

Nothing actually - I've updated the test.